### ISSUE TITLE
initdb: error "/var/lib/postgresql/data" exists but is not empty

### ISSUE THREAD

**Issue by RashiqAzhan** [pgautoupgrade/docker-pgautoupgrade#7]
I'm trying to upgrade from postgres 14 to 15 while retaining data.

When I use the image to `pgautoupgrade/pgautoupgrade:latest` instead of `postgres:14` in the compose file and restart, I get the following errors.

```bash
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
If you want to create a new database system, either remove or empty
he directory "/var/lib/postgresql/data" or run initdb
with an argument other than "/var/lib/postgresql/data".
```

Maybe I'm misunderstanding but shouldn't the data directory be not empty? It should, and does, have data from version 14 that I want migrated to 15. Thank you for any help.

**Comment by justinclift** [comment]
Oh, now that's strange.  It's *supposed* to have moved your old data files to an "old" subdirectory prior to doing that, and it's also supposed to have created a brand new subdirectory to do the `initdb` into (called "new").

https://github.com/pgautoupgrade/docker-pgautoupgrade/blob/585a87d03365fa238915a933c24d1bf68c9af3f6/docker-entrypoint.sh#L377-L402

Which docker image / tag is this happening with?

**Comment by RashiqAzhan** [comment]
I'm using `postgres:14.5` from [here](https://hub.docker.com/_/postgres).

I would also like say I have this warning regarding `psql`.
```bash
WARNING: The supplied POSTGRES_PASSWORD is 100+ characters.

  This will not work if used via PGPASSWORD with "psql".

  https://www.postgresql.org/message-id/flat/E1Rqxp2-0004Qt-PL%40wrigleys.postgresql.org (BUG #6412)
  https://github.com/docker-library/postgres/issues/507

The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
```

Could 100+ character could be a sticking point?

**Comment by RashiqAzhan** [comment]
Thank you so much. I understand this is an open source project I greatly appreciate your contribution.

(Fat fingers, erroneously closed the issue)

**Comment by justinclift** [comment]
No worries at all. :smile:

**Comment by Heziode** [comment]
I also got this error, starting from an `postgres:11-alpine` to `pgautoupgrade/pgautoupgrade:latest` (currently latest: 15-alpine3.8).

I upgrading a [Weblate](https://github.com/WeblateOrg/weblate) database.

Here is my log:

```
PostgreSQL Database directory appears to contain a database; Skipping initialization
******************************************************
Performing PG upgrade on version 11 database files
******************************************************
Moving the old database files prior to pg_upgrade
'/var/lib/postgresql/data/PG_VERSION' -> '/var/lib/postgresql/data/old/PG_VERSION'
'/var/lib/postgresql/data/base' -> '/var/lib/postgresql/data/old/base'
'/var/lib/postgresql/data/global' -> '/var/lib/postgresql/data/old/global'
mv: can't rename '/var/lib/postgresql/data/old': Invalid argument
'/var/lib/postgresql/data/old' -> '/var/lib/postgresql/data/old/old'
'/var/lib/postgresql/data/pg_commit_ts' -> '/var/lib/postgresql/data/old/pg_commit_ts'
'/var/lib/postgresql/data/pg_dynshmem' -> '/var/lib/postgresql/data/old/pg_dynshmem'
'/var/lib/postgresql/data/pg_hba.conf' -> '/var/lib/postgresql/data/old/pg_hba.conf'
'/var/lib/postgresql/data/pg_ident.conf' -> '/var/lib/postgresql/data/old/pg_ident.conf'
'/var/lib/postgresql/data/pg_logical' -> '/var/lib/postgresql/data/old/pg_logical'
'/var/lib/postgresql/data/pg_multixact' -> '/var/lib/postgresql/data/old/pg_multixact'
'/var/lib/postgresql/data/pg_notify' -> '/var/lib/postgresql/data/old/pg_notify'
'/var/lib/postgresql/data/pg_replslot' -> '/var/lib/postgresql/data/old/pg_replslot'
'/var/lib/postgresql/data/pg_serial' -> '/var/lib/postgresql/data/old/pg_serial'
'/var/lib/postgresql/data/pg_snapshots' -> '/var/lib/postgresql/data/old/pg_snapshots'
'/var/lib/postgresql/data/pg_stat' -> '/var/lib/postgresql/data/old/pg_stat'
'/var/lib/postgresql/data/pg_stat_tmp' -> '/var/lib/postgresql/data/old/pg_stat_tmp'
'/var/lib/postgresql/data/pg_subtrans' -> '/var/lib/postgresql/data/old/pg_subtrans'
'/var/lib/postgresql/data/pg_tblspc' -> '/var/lib/postgresql/data/old/pg_tblspc'
'/var/lib/postgresql/data/pg_twophase' -> '/var/lib/postgresql/data/old/pg_twophase'
'/var/lib/postgresql/data/pg_wal' -> '/var/lib/postgresql/data/old/pg_wal'
'/var/lib/postgresql/data/pg_xact' -> '/var/lib/postgresql/data/old/pg_xact'
'/var/lib/postgresql/data/postgresql.auto.conf' -> '/var/lib/postgresql/data/old/postgresql.auto.conf'
'/var/lib/postgresql/data/postgresql.conf' -> '/var/lib/postgresql/data/old/postgresql.conf'
'/var/lib/postgresql/data/postmaster.opts' -> '/var/lib/postgresql/data/old/postmaster.opts'
PostgreSQL 11 database files found, upgrading to PostgreSQL 15
Initialising PostgreSQL 15 data directory
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
fixing permissions on existing directory /var/lib/postgresql/data/new ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... UTC
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
syncing data to disk ... ok
Success. You can now start the database server using:
    /usr/local/bin/pg_ctl -D /var/lib/postgresql/data/new/ -l logfile start
initdb: warning: enabling "trust" authentication for local connections
initdb: hint: You can change this by editing pg_hba.conf or using the option -A, or --auth-local and --auth-host, the next time you run initdb.
Performing Consistency Checks
-----------------------------
Checking cluster versions                                   ok
connection to server on socket "/var/lib/postgresql/.s.PGSQL.50432" failed: FATAL:  role "postgres" does not exist
could not connect to source postmaster started with the command:
"/usr/local-pg11/bin/pg_ctl" -w -l "/var/lib/postgresql/data/new/pg_upgrade_output.d/20230806T174251.151/log/pg_upgrade_server.log" -D "/var/lib/postgresql/data/old" -o "-p 50432 -b  -c listen_addresses='' -c unix_socket_permissions=0700 -c unix_socket_directories='/var/lib/postgresql'" start
Failure, exiting
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
Data page checksums are disabled.
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".
Data page checksums are disabled.
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
```

And the corresponding DB service into docker-compose:

```yaml
  database:
    # image: postgres:11-alpine
    image: pgautoupgrade/pgautoupgrade:latest
    container_name: weblate_database
    env_file:
      - ./environment
    volumes:
      - postgres-data:/var/lib/postgresql/data
    restart: always
    networks:
      - backend
```

environment contains Postgres env (I set fake data, but my password use special chars like **@**):

```env
POSTGRES_PASSWORD=p@ssw0rd
POSTGRES_USER=weblate
POSTGRES_DATABASE=weblate
POSTGRES_HOST=database
POSTGRES_PORT=
```

**Comment by justinclift** [comment]
Thanks for the bug report.  Looking at the log output there, this line is standing out:

```
connection to server on socket "/var/lib/postgresql/.s.PGSQL.50432" failed: FATAL:
 role "postgres" does not exist
```

That sounds like your PostgreSQL database uses something other than the "postgres" user as the superuser.  Probably `weblate` by looking at your environment variables. :smile:

Co-incidentally, the recent `dev` tagged image has some support for that.  If you try it again, but using the `dev` tagged image there's a decent chance it'll work.

Would you be ok to try that out and let us know if it does indeed do the job?

If it does, we can promote the `dev` tag to `latest` (etc) so others in the same situation have things automatically work for them too. :smile:

**Comment by Heziode** [comment]
I tested with `dev` tag, it works now!

**Comment by justinclift** [comment]
Excellent!  I've pushed a rebuilt `dev` image to `0.5.0`, `latest`, and `15-alpine3.8` tags.

That should fix this problem for others using a similar setup. :smile:

**Comment by eurekin** [comment]
I have a very similar case. I had the same error as this issue description. Found the culprit in my case, logs:

```
postgres-postgres-1    | ------------------------------------
postgres-postgres-1    | 2023-11-19 22:48:45.170 CET [29] FATAL:  lock file "postmaster.pid" already exists
postgres-postgres-1    | 2023-11-19 22:48:45.170 CET [29] HINT:  Is another postmaster (PID 1) running in data directory "/var/lib/postgresql/data/old"?
```

After that, postgres wanted to start, but the data directory contained new and old subdirs instead of typical data. After restoring the old content, removing the postmaster.pid it proceeded succesfully without any more issues.

I'm assuming the issue arose due to my auto-upgrade setup. Is watchtower recommended for upgrading the postgres? What could be the cause for dangling pidlock file?

**Comment by jjb** [comment]
Hi, thanks so much for this awesome project

I'm getting this error

```
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
```

here's my docker compose

```yaml
  postgres:
    image: 'pgautoupgrade/pgautoupgrade:16-alpine'
    environment:
      POSTGRES_USER: 'redacted'
      POSTGRES_PASSWORD: 'redacted'
      POSTGRES_DB: 'redacted'
      PGDATA: '/var/lib/postgresql/data/pgdata'
    volumes:
      - 'postgres:/var/lib/postgresql/data'
    ports:
      - '5432'
```

before the upgrade, image was `'postgres:12-alpine'`

any ideas?

**Comment by justinclift** [comment]
Hmmm, that *sounds* like the `initdb` command the script runs is somehow not being pointed to `/var/lib/postgresql/data/new`.  That's a temporary directory created by the script before hand just for initdb purposes.

The way I tend to debug stuff like this is by starting the container (with this image and the PG data mounted), but tell the script to stop *just before* the upgrade runs.  Then I manually run the lines of the upgrade script 1 by 1 to see where it breaks, and why.

That whole "tell it to stop just before the upgrade script runs" is done by defining the environment variable `PGAUTO_DEVEL=before` before starting the container.  I normally do it from the command line, but you could do the same thing using the docker compose.

If you're up for investigating, that'd be the approach to use.  Note, it's not something you'd want to be debugging live on your production database, as the database will be unavailable to other applications meanwhile.  ie if you're up for it, then do it in a copy somewhere where it won't affect things

Thoughts? :smile:

**Comment by jjb** [comment]
will work on that

**Comment by justinclift** [comment]
For reference, this is the code to manually step through when trying to figure out what's going wrong:

https://github.com/pgautoupgrade/docker-pgautoupgrade/blob/54179ffc234a5a095bf7e7e5a52f0eb52a7f9912/docker-entrypoint.sh#L362-L553

And yep, it's copiously commented and tries to be as straight forward as possible.  ie tries to be easy to see what it's doing at all times

**Comment by jjb** [comment]
thanks!

so, the place where i have the problem is a hosted environment that doesn't have great logs and such. but i can push whatever i want there and then ssh in while it's running.

i'm not the greatest at docker - here's how i'm trying to deploy without doing any upgrades: https://github.com/jjb/docker-pgautoupgrade/commit/abbf1e31eafe39f4afabc124095bb5773e6810e7

that failed, without any logs (long story). any ideas?

**Comment by justinclift** [comment]
Ahhh.  An important concept with Docker is that a container only runs while there's a main process running.  If that main process ends, then the docker container will exit thinking it's work has been done.

By commenting out those three lines, you've most likely stopped the PostgreSQL (PG) server from launching... so there's nothing left to run and the container just exits.

To work around that, add a never ending while loop to the end of that docker entry point script:

```
while :; do
    sleep 5
done
```

That'll leave the container running, then with another instance of your terminal program you can jump into the running container:

```
$ docker exec -it CONTAINER_NAME /bin/bash
```

From there, you can switch to the `postgres` user and prod things as needed.

Does that help? :smile:

**Comment by jjb** [comment]
never mind, i see the entrypoint needs to run... will figure out a better way to NOOP it

**Comment by jjb** [comment]
haha posted that at the same time your message came in

thanks for the suggestion! in my case, pg does need to come up so healtchchecks are good, otherwise i can't ssh in. so i just need to NOOP the upgrade and allow postgres to run. looking making a clean short circuit now.

**Comment by justinclift** [comment]
Cool, all good. :smile:

**Comment by jjb** [comment]
🤷  https://github.com/jjb/docker-pgautoupgrade/commit/24b809bcd873cf951303460d15dc0274f3c80fc8

**Comment by jjb** [comment]
> "root" execution of the PostgreSQL server is not permitted.
> The server must be started under an unprivileged user ID to prevent
> possible system security compromise.  See the documentation for
> more information on how to properly start the server.

**Comment by justinclift** [comment]
Hmmm, if you want to have PostgreSQL run without the upgrade happening, you could just comment out the whole upgrade piece.  Or you could muck around with the test that it uses for determining if the upgrade should happen:

https://github.com/jjb/docker-pgautoupgrade/blob/24b809bcd873cf951303460d15dc0274f3c80fc8/docker-entrypoint.sh#L375

Changing that line to something like this would stop the upgrade from automatically running too:

```
if [ "one" != "two" ]; then
```

^^^ that'll always fail, so the stuff inside the `then` clause won't run.

**Comment by justinclift** [comment]
```
"root" execution of the PostgreSQL server is not permitted.
```

Is that when trying to manually start it after you've jumped into a running container?  If so, then switch to the `postgres` user first (`su - postgres`).

**Comment by jjb** [comment]
trying this at the end of the entrypoint script

```bash
exec su-exec postgres "$@"
```

**Comment by justinclift** [comment]
Hmmm, that doesn't sound like the right approach (to me).

---

Oh, there might be a conceptual misunderstanding.

For the upgrade pieces to run, it needs to happen when the PostgreSQL container itself is running... but the PostgreSQL server software inside the container has to be shut down.

That's why the upgrade scripting check and commands are before the entry point script starts PostgreSQL.  That way, the upgrade is all taken care of and done before PostgreSQL starts.

Does that make sense, or am I misunderstanding the approach you're taking? :smile:

**Comment by jjb** [comment]
sure, that makes sense

but, can't i still hack around on the running instance? (this is all in a throw-away environment) either by first shutting down pg, or just not caring and plowing over files to test the commands?

**Comment by jjb** [comment]
> postgres: could not access the server configuration file "/var/lib/postgresql/data/pgdata/postgresql.conf": No such file or directory

i'll try your one=two approach

**Comment by justinclift** [comment]
Sure, you definitely can.  Try running the docker container with that `PGAUTO_DEVEL` environment variable defined.

That'll leave the docker container running, before the upgrade script runs.

You can then jump into the container (from another terminal session) using the `docker exec` command, and if you switch to the `postgres` user you can start the database (`pg_ctl start`) and/or try manually running the upgrade commands.

**Comment by jjb** [comment]
in my environment, i need a passing pg healthcheck before i can SSH in

**Comment by jjb** [comment]
<img width="473" alt="image" src="https://github.com/pgautoupgrade/docker-pgautoupgrade/assets/6097/37dd1858-c256-404e-81c7-4df265f45d3d">

> initdb: error: directory "/var/lib/postgresql/data/pgdata" exists but is not empty
> initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data/pgdata" or run initdb with an argument other than "/var/lib/postgresql/data/pgdata".

**Comment by justinclift** [comment]
The `Makefile` in the same repo might be good reference info too:

https://github.com/jjb/docker-pgautoupgrade/blob/24b809bcd873cf951303460d15dc0274f3c80fc8/Makefile

Still like the `before:` target shows how to manually launch the container and tell the script to stop at one of the defined break points:

```
$ docker run --name pgauto -it --rm \
    --mount type=bind,source=postgres,target=/var/lib/postgresql/data \
    -e POSTGRES_PASSWORD=password \
    -e PGAUTO_DEVEL=before \
    NAME_OF_IMAGE_TO_RUN
```

You might want to leave off the `--rm` option, as that tells Docker to nuke the container when it exits.  Leaving it hanging around means you can start it up again which is useful sometimes.

**Comment by justinclift** [comment]
> in my environment, i need a passing pg healthcheck before i can SSH in

Ahhhh.  Oh, now that all makes sense.

Ahhh, I have an idea.  Gimme a minute to write it... :smile:

**Comment by justinclift** [comment]
Ok, so the concept I'm thinking of here is mostly working around that health check requirement. :grin:

So, the things we need to have happen are:

a) The container runs, starting PostgreSQL normally *without* attempting an upgrade.

Your change of `if [ false ]; then` should accomplish that.

b) We need to shut down the PostgreSQL software inside the container *after* you've ssh-d in.

If we do that though, then the main running loop of the program will be over, and the container will exit.  But, we already know how to fix that using an endless while loop:

```
while :; do
    sleep 5
done
```

Add add one of those to the very end of the entry point script, so after we tell PostgreSQL to stop then the container keeps on running.

c) Now you should be able to jump into the container (`docker exec`), switch to the `postgres` user, and then try to figure out why the upgrade is breaking.

---

So, that's only two small changes needed to the stock entry point script, and that should get you in the right spot.

How does that sound? :smile:

**Comment by jjb** [comment]
sounds good, thanks! but i think maybe you missed the error message i got when i tried my hack - see the text under the image here https://github.com/pgautoupgrade/docker-pgautoupgrade/issues/7#issuecomment-2000638399

**Comment by jjb** [comment]
so the `flase` change doesn't NOOP the whole script, this stuff still runs, i don't understand the system enough to know why

```bash
	if [ "$1" = 'postgres' ] && ! _pg_want_help "$@"; then
		docker_setup_env
		# setup data directories and permissions (when run as root)
		docker_create_db_directories
```

**Comment by justinclift** [comment]
Ahhh.  Hmmm, that error message isn't making sense to me if that's the only change to the script.

---

Oh, it sounds like it's erroring out before it even gets to the upgrade piece.

Just as an experiment, I'd try passing in an empty directory for your PostgreSQL data directory and see if that script still errors out or if the initdb piece runs.

It could turn out to be something else unexpected is going on.

**Comment by jjb** [comment]
okay, trying this

```bash
_main() {
	mkdir /tmp/my-empty-dir
```

```yaml
    environment:
      PGDATA: '/tmp/my-empty-dir'
    volumes:
      - 'postgres:/tmp/my-empty-dir'
```

**Comment by justinclift** [comment]
Hmmm, you might want to leave off the `volumes` line (and the one below it), otherwise I think it's going to still mount your local PG data into the container at that location.

**Comment by jjb** [comment]
okay, just nuke the whole `volumes` section?

**Comment by justinclift** [comment]
Yep, for testing purposes, just to see if the script runs ok with it gone. :smile:

**Comment by justinclift** [comment]
It looks like it's getting towards 8pm your time, and I'm starting to need sleep pretty badly myself.

If you're up for it, we can schedule a time to investigate this properly over Skype or something when we're both fully awake? :smile:

**Comment by jjb** [comment]
wow, thanks for the offer - although i fear the turnaround time on each experiment will be too long, so maybe we should get me a working shell first. you can find my email off of my website linked here https://github.com/jjb

**Comment by jjb** [comment]
progress!  and now i realized i should be using my existing pg version image 🤦 

```
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /tmp/my-empty-dir ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... UTC
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
syncing data to disk ... ok


initdb: warning: enabling "trust" authentication for local connections
initdb: hint: You can change this by editing pg_hba.conf or using the option -A, or --auth-local and --auth-host, the next time you run initdb.
Success. You can now start the database server using:

    pg_ctl -D /tmp/my-empty-dir -l logfile start

waiting for server to start....2024-03-15 23:58:01.233 UTC [39] LOG:  starting PostgreSQL 16.2 on x86_64-pc-linux-musl, compiled by gcc (Alpine 13.2.1_git20231014) 13.2.1 20231014, 64-bit
2024-03-15 23:58:01.237 UTC [39] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2024-03-15 23:58:01.249 UTC [42] LOG:  database system was shut down at 2024-03-15 23:57:57 UTC
2024-03-15 23:58:01.257 UTC [39] LOG:  database system is ready to accept connections
 done
server started
CREATE DATABASE


/usr/local/bin/docker-entrypoint.sh: ignoring /docker-entrypoint-initdb.d/*

waiting for server to shut down....2024-03-15 23:58:01.394 UTC [39] LOG:  received fast shutdown request
2024-03-15 23:58:01.399 UTC [39] LOG:  aborting any active transactions
2024-03-15 23:58:01.401 UTC [39] LOG:  background worker "logical replication launcher" (PID 45) exited with exit code 1
2024-03-15 23:58:01.401 UTC [40] LOG:  shutting down
2024-03-15 23:58:01.405 UTC [40] LOG:  checkpoint starting: shutdown immediate
.2024-03-15 23:58:02.454 UTC [40] LOG:  checkpoint complete: wrote 935 buffers (5.7%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.016 s, sync=1.016 s, total=1.053 s; sync files=301, longest=0.014 s, average=0.004 s; distance=4327 kB, estimate=4327 kB; lsn=0/19495A8, redo lsn=0/19495A8
2024-03-15 23:58:02.460 UTC [39] LOG:  database system is shut down
 done
server stopped

PostgreSQL init process complete; ready for start up.

************************************
PostgreSQL data directory: /tmp/my-empty-dir
************************************
****************************************************************************
Unrecognised version of PostgreSQL database files found, aborting completely
****************************************************************************

```

**Comment by jjb** [comment]
my existing version: 12 🫠

**Comment by justinclift** [comment]
Ok, well that gets it a bit closer. :smile:

I've just emailed you (`Ping...` as subject), so we can figure things out when you're up for it. :grin:

**Comment by justinclift** [comment]
Heh Heh Heh.  12 isn't the worst I've seen.  I started this project due to needing to upgrade a bunch of 9.something production systems. :grin:

**Comment by jjb** [comment]
🫠 isn't because it's old but because your upgrade image doesn't support it... so i can't start up my disabled fork for 12.. right? but maybe i can simply hack your build scripts easily enough

**Comment by justinclift** [comment]
Heh Heh Heh.  Is that `false` check still in place?

The pgautoupgrade container should definitely support upgrading from PG 12.

**Comment by jjb** [comment]
seems to be building just fine (will be half an hour because building AMD on my ARM mac)

**Comment by jjb** [comment]
> The pgautoupgrade container should definitely support upgrading from PG 12.

right but the makeful doesn't support building `PGTARGET=12`

but changing up the invocation was trivial, so i think we are off to the races

**Comment by justinclift** [comment]
Ahhhh.  You're meaning going *to* PG 12.  Got it.  Yeah, out by one version.

You're right, you should be able to hack that into the Dockerfile and script without too much hassle. :smile:

**Comment by jjb** [comment]
i think it made it past this

```
initdb: warning: enabling "trust" authentication for local connections
initdb: hint: You can change this by editing pg_hba.conf or using the option -A, or --auth-local and --auth-host, the next time you run initdb.
```

and then it died after this

```
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /tmp/my-empty-dir ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... UTC
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
syncing data to disk ... ok


Success. You can now start the database server using:

    pg_ctl -D /tmp/my-empty-dir -l logfile start

initdb: warning: enabling "trust" authentication for local connections
initdb: hint: You can change this by editing pg_hba.conf or using the option -A, or --auth-local and --auth-host, the next time you run initdb.
waiting for server to start....2024-03-16 00:13:11.133 UTC [39] LOG:  starting PostgreSQL 16.2 on x86_64-pc-linux-musl, compiled by gcc (Alpine 13.2.1_git20231014) 13.2.1 20231014, 64-bit
2024-03-16 00:13:11.136 UTC [39] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2024-03-16 00:13:11.149 UTC [42] LOG:  database system was shut down at 2024-03-16 00:13:07 UTC
2024-03-16 00:13:11.156 UTC [39] LOG:  database system is ready to accept connections
 done
server started
CREATE DATABASE


/usr/local/bin/docker-entrypoint.sh: ignoring /docker-entrypoint-initdb.d/*

waiting for server to shut down....2024-03-16 00:13:11.283 UTC [39] LOG:  received fast shutdown request
2024-03-16 00:13:11.288 UTC [39] LOG:  aborting any active transactions
2024-03-16 00:13:11.289 UTC [39] LOG:  background worker "logical replication launcher" (PID 45) exited with exit code 1
2024-03-16 00:13:11.291 UTC [40] LOG:  shutting down
2024-03-16 00:13:11.294 UTC [40] LOG:  checkpoint starting: shutdown immediate
.2024-03-16 00:13:12.346 UTC [40] LOG:  checkpoint complete: wrote 935 buffers (5.7%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.016 s, sync=1.016 s, total=1.056 s; sync files=301, longest=0.012 s, average=0.004 s; distance=4327 kB, estimate=4327 kB; lsn=0/19495A8, redo lsn=0/19495A8
2024-03-16 00:13:12.352 UTC [39] LOG:  database system is shut down
 done
server stopped

PostgreSQL init process complete; ready for start up.

************************************
PostgreSQL data directory: /tmp/my-empty-dir
************************************
****************************************************************************
Unrecognised version of PostgreSQL database files found, aborting completely
****************************************************************************
```

**Comment by justinclift** [comment]
Interesting.  That `Unrecognised version of PostgreSQL database files found, aborting completely` should only happen if there's an "unknown" version of PostgreSQL data files in the data directory.  In theory.

k, plan b, lets add a version of that volume line back to your compose file, and give it an empty directory on your local system to mount:

```
    environment:
      PGDATA: '/var/lib/postgresql/data'
    volumes:
      - '/tmp/pgtemp1:/var/lib/postgresql/data'
```

That should create an empty directory `/tmp/pgtemp1` on your local system, and mount that inside the container at `/var/lib/postgresql/data`.

When the container starts, it should see that directory is empty and initialise it with a blank PostgreSQL database.

*If* the rest of the container startup process errors out again, you can look at the `/tmp/pgtemp1` directory on your local system to see WTF is in the files there. :wink:

There should be a `PG_VERSION` file (plain text`) in the directory after it's initialised, that says what version of PostgreSQL did the initialisation.

**Comment by jjb** [comment]
```
PostgreSQL Database directory appears to contain a database; Skipping initialization

************************************
PostgreSQL data directory: /var/lib/postgresql/data
************************************
****************************************************************************
Unrecognised version of PostgreSQL database files found, aborting completely
****************************************************************************
```

**Comment by jjb** [comment]
i knew i was flying too close to the sun thinking i could use `false` in shell scripting. my text editor highlights it a different color! why!

```
postgres-6d5984cd68-7fgng:/# if [ "one" == "two" ]; then echo yes; fi
postgres-6d5984cd68-7fgng:/# if [ false ]; then echo yes; fi
yes
```

**Comment by jjb** [comment]
okay, in additinon to the above facepalm, i also realized i wasn't deploying my patched version of your code but was building main.

now i'm back on track possibly getting better info. i got a build with all this glorious log output. then i put i more logging and for some reason the next run didn't have any output. still trying to figure out if my fault or something else, but want to safe this here for posterity. the interesting things are 

* in the middle `mv: can't rename '/var/lib/postgresql/data/pgdata/old': Invalid argument` - i put in some logging here to try to figure out what it's trying to move, but that didn't work... yet
* at the end
  ```
  2024-03-16 02:20:41.927 UTC [30] FATAL:  lock file "postmaster.pid" already exists
  2024-03-16 02:20:41.927 UTC [30] HINT:  Is another postmaster (PID 1) running in data directory "/var/lib/postgresql/data/pgdata/old"?
  ```

 
```
PostgreSQL Database directory appears to contain a database; Skipping initialization

************************************
PostgreSQL data directory: /var/lib/postgresql/data/pgdata
************************************
*******************************************************************************************
Performing PG upgrade on version 12 database files.  Upgrading to version 16
*******************************************************************************************
----------------------------------------------------------------------
Checking for left over artifacts from a failed previous autoupgrade...
----------------------------------------------------------------------
-------------------------------------------------------------------------------
No artifacts found from a failed previous autoupgrade.  Continuing the process.
-------------------------------------------------------------------------------
---------------------------------------
Creating OLD temporary directory /var/lib/postgresql/data/pgdata/old
---------------------------------------
--------------------------------------------
Creating OLD temporary directory is complete
--------------------------------------------
-------------------------------------------------------
Moving existing data files into OLD temporary directory
-------------------------------------------------------
'/var/lib/postgresql/data/pgdata/PG_VERSION' -> '/var/lib/postgresql/data/pgdata/old/PG_VERSION'
'/var/lib/postgresql/data/pgdata/base' -> '/var/lib/postgresql/data/pgdata/old/base'
mv: can't rename '/var/lib/postgresql/data/pgdata/old': Invalid argument
'/var/lib/postgresql/data/pgdata/global' -> '/var/lib/postgresql/data/pgdata/old/global'
'/var/lib/postgresql/data/pgdata/old' -> '/var/lib/postgresql/data/pgdata/old/old'
'/var/lib/postgresql/data/pgdata/pg_commit_ts' -> '/var/lib/postgresql/data/pgdata/old/pg_commit_ts'
'/var/lib/postgresql/data/pgdata/pg_dynshmem' -> '/var/lib/postgresql/data/pgdata/old/pg_dynshmem'
'/var/lib/postgresql/data/pgdata/pg_hba.conf' -> '/var/lib/postgresql/data/pgdata/old/pg_hba.conf'
'/var/lib/postgresql/data/pgdata/pg_ident.conf' -> '/var/lib/postgresql/data/pgdata/old/pg_ident.conf'
'/var/lib/postgresql/data/pgdata/pg_logical' -> '/var/lib/postgresql/data/pgdata/old/pg_logical'
'/var/lib/postgresql/data/pgdata/pg_multixact' -> '/var/lib/postgresql/data/pgdata/old/pg_multixact'
'/var/lib/postgresql/data/pgdata/pg_notify' -> '/var/lib/postgresql/data/pgdata/old/pg_notify'
'/var/lib/postgresql/data/pgdata/pg_replslot' -> '/var/lib/postgresql/data/pgdata/old/pg_replslot'
'/var/lib/postgresql/data/pgdata/pg_serial' -> '/var/lib/postgresql/data/pgdata/old/pg_serial'
'/var/lib/postgresql/data/pgdata/pg_snapshots' -> '/var/lib/postgresql/data/pgdata/old/pg_snapshots'
'/var/lib/postgresql/data/pgdata/pg_stat' -> '/var/lib/postgresql/data/pgdata/old/pg_stat'
'/var/lib/postgresql/data/pgdata/pg_stat_tmp' -> '/var/lib/postgresql/data/pgdata/old/pg_stat_tmp'
'/var/lib/postgresql/data/pgdata/pg_subtrans' -> '/var/lib/postgresql/data/pgdata/old/pg_subtrans'
'/var/lib/postgresql/data/pgdata/pg_tblspc' -> '/var/lib/postgresql/data/pgdata/old/pg_tblspc'
'/var/lib/postgresql/data/pgdata/pg_twophase' -> '/var/lib/postgresql/data/pgdata/old/pg_twophase'
'/var/lib/postgresql/data/pgdata/pg_wal' -> '/var/lib/postgresql/data/pgdata/old/pg_wal'
'/var/lib/postgresql/data/pgdata/pg_xact' -> '/var/lib/postgresql/data/pgdata/old/pg_xact'
'/var/lib/postgresql/data/pgdata/postgresql.auto.conf' -> '/var/lib/postgresql/data/pgdata/old/postgresql.auto.conf'
'/var/lib/postgresql/data/pgdata/postgresql.conf' -> '/var/lib/postgresql/data/pgdata/old/postgresql.conf'
'/var/lib/postgresql/data/pgdata/postmaster.opts' -> '/var/lib/postgresql/data/pgdata/old/postmaster.opts'
'/var/lib/postgresql/data/pgdata/postmaster.pid' -> '/var/lib/postgresql/data/pgdata/old/postmaster.pid'
-------------------------------------------------------------------
Moving existing data files into OLD temporary directory is complete
-------------------------------------------------------------------
---------------------------------------
Creating NEW temporary directory /var/lib/postgresql/data/pgdata/new
---------------------------------------
--------------------------------------------
Creating NEW temporary directory is complete
--------------------------------------------
-----------------------------------------------------
Changing permissions of temporary directories to 0700
-----------------------------------------------------
---------------------------------------------------------
Changing permissions of temporary directories is complete
---------------------------------------------------------
------------------------------------
Determining our own initdb arguments
------------------------------------
2024-03-16 02:20:41.927 UTC [30] FATAL:  lock file "postmaster.pid" already exists
2024-03-16 02:20:41.927 UTC [30] HINT:  Is another postmaster (PID 1) running in data directory "/var/lib/postgresql/data/pgdata/old"?
```

**Comment by jjb** [comment]
In the light of morning I _quickly_ created a minimal local reproduction!

```yaml
version: '3'

services:
  postgres:
    image: 'pgautoupgrade/pgautoupgrade:15-alpine'
    platform: 'linux/amd64'
    environment:
      POSTGRES_PASSWORD: 'password'
      POSTGRES_DB: 'app'
      PGDATA: '/var/lib/postgresql/data/pgdata'
    volumes:
      - 'postgres:/var/lib/postgresql/data'
    ports:
      - '5432'
```

```shell
docker-compose up postgres
```

```
 | PostgreSQL Database directory appears to contain a database; Skipping initialization
 |
 | ************************************
 | PostgreSQL data directory: /var/lib/postgresql/data/pgdata
 | ************************************
 | ****************************************************************************
 | Unrecognised version of PostgreSQL database files found, aborting completely
 | ****************************************************************************
exited with code 9
```

**Comment by jjb** [comment]
okay, now realizing that isn't actually a standalone reproduction, requires the volume having something in it first. standby...

**Comment by justinclift** [comment]
With your PostgreSQL data directory, what version of PostgreSQL is it upgrading from?

I should update the script in the main repository to support PG 12 as the target.  That would probably help your situation out a lot. :smile:

**Comment by justinclift** [comment]
I've just added support for migrating to PG 12 (7ffe5e7ae59419e09903d4d01614992da00a5708), and pushed a new Docker image with the tag `12-dev` to the Docker Hub repository.

If you try that docker image out, it *might* just all work out automatically. :smile:

**Comment by justinclift** [comment]
As an update for anyone keeping an eye on this issue, the problem that @jjb was hitting was really two separate things.  Both now fixed (https://github.com/pgautoupgrade/docker-pgautoupgrade/commit/ba99db1a6398932999146059d90c5be23bb0375e) and (https://github.com/pgautoupgrade/docker-pgautoupgrade/commit/c869388e73312c842126deff8cfb57b20d85c8b6).

@eurekin From the description of the problem you hit, the 2nd of those might have fixed things for you as well. :smile:

**Comment by FireMasterK** [comment]
<details>

```
PostgreSQL Database directory appears to contain a database; Skipping initialization

************************************
PostgreSQL data directory: /var/lib/postgresql/data
************************************
*******************************************************************************************
Performing PG upgrade on version 15 database files.  Upgrading to version 16
*******************************************************************************************
----------------------------------------------------------------------
Checking for left over artifacts from a failed previous autoupgrade...
----------------------------------------------------------------------
-------------------------------------------------------------------------------
No artifacts found from a failed previous autoupgrade.  Continuing the process.
-------------------------------------------------------------------------------
---------------------------------------
Creating OLD temporary directory /var/lib/postgresql/data/old
---------------------------------------
--------------------------------------------
Creating OLD temporary directory is complete
--------------------------------------------
-------------------------------------------------------
Moving existing data files into OLD temporary directory
-------------------------------------------------------
'/var/lib/postgresql/data/PG_VERSION' -> '/var/lib/postgresql/data/old/PG_VERSION'
'/var/lib/postgresql/data/base' -> '/var/lib/postgresql/data/old/base'
'/var/lib/postgresql/data/core' -> '/var/lib/postgresql/data/old/core'
'/var/lib/postgresql/data/global' -> '/var/lib/postgresql/data/old/global'
mv: can't rename '/var/lib/postgresql/data/old': Invalid argument
'/var/lib/postgresql/data/old' -> '/var/lib/postgresql/data/old/old'
'/var/lib/postgresql/data/pg_commit_ts' -> '/var/lib/postgresql/data/old/pg_commit_ts'
'/var/lib/postgresql/data/pg_dynshmem' -> '/var/lib/postgresql/data/old/pg_dynshmem'
'/var/lib/postgresql/data/pg_hba.conf' -> '/var/lib/postgresql/data/old/pg_hba.conf'
'/var/lib/postgresql/data/pg_ident.conf' -> '/var/lib/postgresql/data/old/pg_ident.conf'
'/var/lib/postgresql/data/pg_logical' -> '/var/lib/postgresql/data/old/pg_logical'
'/var/lib/postgresql/data/pg_multixact' -> '/var/lib/postgresql/data/old/pg_multixact'
'/var/lib/postgresql/data/pg_notify' -> '/var/lib/postgresql/data/old/pg_notify'
'/var/lib/postgresql/data/pg_replslot' -> '/var/lib/postgresql/data/old/pg_replslot'
'/var/lib/postgresql/data/pg_serial' -> '/var/lib/postgresql/data/old/pg_serial'
'/var/lib/postgresql/data/pg_snapshots' -> '/var/lib/postgresql/data/old/pg_snapshots'
'/var/lib/postgresql/data/pg_stat' -> '/var/lib/postgresql/data/old/pg_stat'
'/var/lib/postgresql/data/pg_stat_tmp' -> '/var/lib/postgresql/data/old/pg_stat_tmp'
'/var/lib/postgresql/data/pg_subtrans' -> '/var/lib/postgresql/data/old/pg_subtrans'
'/var/lib/postgresql/data/pg_tblspc' -> '/var/lib/postgresql/data/old/pg_tblspc'
'/var/lib/postgresql/data/pg_twophase' -> '/var/lib/postgresql/data/old/pg_twophase'
'/var/lib/postgresql/data/pg_wal' -> '/var/lib/postgresql/data/old/pg_wal'
'/var/lib/postgresql/data/pg_xact' -> '/var/lib/postgresql/data/old/pg_xact'
'/var/lib/postgresql/data/postgresql.auto.conf' -> '/var/lib/postgresql/data/old/postgresql.auto.conf'
'/var/lib/postgresql/data/postgresql.conf' -> '/var/lib/postgresql/data/old/postgresql.conf'
'/var/lib/postgresql/data/postmaster.opts' -> '/var/lib/postgresql/data/old/postmaster.opts'
-------------------------------------------------------------------
Moving existing data files into OLD temporary directory is complete
-------------------------------------------------------------------
---------------------------------------
Creating NEW temporary directory /var/lib/postgresql/data/new
---------------------------------------
--------------------------------------------
Creating NEW temporary directory is complete
--------------------------------------------
-----------------------------------------------------
Changing permissions of temporary directories to 0700
-----------------------------------------------------
---------------------------------------------------------
Changing permissions of temporary directories is complete
---------------------------------------------------------
------------------------------------------------------------------------------
Using initdb arguments passed in from the environment: "-E UTF8"
------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------
Old database using collation settings: '"-E UTF8"'.  Initialising new database with those settings too
--------------------------------------------------------------------------------------------------------------------
Initialising PostgreSQL 16 data directory
initdb: error: too many command-line arguments (first is "UTF8"")
initdb: hint: Try "initdb --help" for more information.
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with this locale configuration:
  provider:    libc
  LC_COLLATE:  C
  LC_CTYPE:    C
  LC_MESSAGES: en_US.utf8
  LC_MONETARY: en_US.utf8
  LC_NUMERIC:  en_US.utf8
  LC_TIME:     en_US.utf8
The default text search configuration will be set to "english".

Data page checksums are disabled.
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".

The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with this locale configuration:
  provider:    libc
  LC_COLLATE:  C
  LC_CTYPE:    C
  LC_MESSAGES: en_US.utf8
  LC_MONETARY: en_US.utf8
  LC_NUMERIC:  en_US.utf8
  LC_TIME:     en_US.utf8
The default text search configuration will be set to "english".

Data page checksums are disabled.

initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with this locale configuration:
  provider:    libc
  LC_COLLATE:  C
  LC_CTYPE:    C
  LC_MESSAGES: en_US.utf8
initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
  LC_MONETARY: en_US.utf8
  LC_NUMERIC:  en_US.utf8
  LC_TIME:     en_US.utf8
The default text search configuration will be set to "english".

Data page checksums are disabled.
```
</details>

I got this error when using `pgautoupgrade/pgautoupgrade:16-alpine`
```
    postgres:
        image: pgautoupgrade/pgautoupgrade:16-alpine
        restart: unless-stopped
        environment:
            - POSTGRES_USER=appservice
            - POSTGRES_PASSWORD=changeme
            - LC_COLLATE=C
            - LC_CTYPE=C
            - POSTGRES_INITDB_ARGS="-E UTF8"
        volumes:
            - pg_db:/var/lib/postgresql/data
        container_name: postgres
        shm_size: '1gb'
```

The volume seems to have a `old` and `new` directory after running it, but `new` is empty.

**Comment by uwekoenig** [comment]
I can confirm the issue with `POSTGRES_INITDB_ARGS` mentioned by @FireMasterK 

Wanted to upgrade from image `postgres:12-alpine` to `pgautoupgrade/pgautoupgrade:15-alpine` and had to remove the line with the arguments to run the database migration.

```
  db:
#    image: postgres:12-alpine
    image: pgautoupgrade/pgautoupgrade:15-alpine
    container_name: xwiki-postgres-db
    volumes:
      - postgres-data:/var/lib/postgresql/data
    environment:
      - POSTGRES_ROOT_PASSWORD=xwiki
      - POSTGRES_PASSWORD=xwiki
      - POSTGRES_USER=xwiki
      - POSTGRES_DB=xwiki
#      - POSTGRES_INITDB_ARGS="--encoding=UTF8"
    networks:
      - bridge
    restart: unless-stopped
```

**Comment by ElDavoo** [comment]
For some reason, in my docker compose I had the same problem, but not when I ran the container directly with docker run.

**Comment by xd003** [comment]
I am facing this very error from when upgrading from postgres 16 to 17 with the One shot mode. 
Command - `docker run --name pgauto -it --mount type=bind,source=/opt/containers/postgres,target=/var/lib/postgresql/data -e POSTGRES_USER=XXXX -e POSTGRES_PASSWORD=XXXXXXXX -e PGAUTO_ONESHOT=yes pgautoupgrade/pgautoupgrade:17-bookworm`

Console Output
```
PostgreSQL Database directory appears to contain a database; Skipping initialization

************************************
PostgreSQL data directory: /var/lib/postgresql/data
************************************
Creating upgrade lock file at /var/lib/postgresql/data/upgrade_in_progress.lock
*******************************************************************************************
Performing PG upgrade on version 16 database files.  Upgrading to version 17
*******************************************************************************************
----------------------------------------------------------------------
Checking for left over artifacts from a failed previous autoupgrade...
----------------------------------------------------------------------
-------------------------------------------------------------------------------
No artifacts found from a failed previous autoupgrade.  Continuing the process.
-------------------------------------------------------------------------------
---------------------------------------
Creating OLD temporary directory /var/lib/postgresql/data/old
---------------------------------------
--------------------------------------------
Creating OLD temporary directory is complete
--------------------------------------------
-------------------------------------------------------
Moving existing data files into OLD temporary directory
-------------------------------------------------------
renamed '/var/lib/postgresql/data/base' -> '/var/lib/postgresql/data/old/base'
renamed '/var/lib/postgresql/data/global' -> '/var/lib/postgresql/data/old/global'
mv: cannot move '/var/lib/postgresql/data/old' to a subdirectory of itself, '/var/lib/postgresql/data/old/old'
renamed '/var/lib/postgresql/data/pg_commit_ts' -> '/var/lib/postgresql/data/old/pg_commit_ts'
renamed '/var/lib/postgresql/data/pg_dynshmem' -> '/var/lib/postgresql/data/old/pg_dynshmem'
renamed '/var/lib/postgresql/data/pg_hba.conf' -> '/var/lib/postgresql/data/old/pg_hba.conf'
renamed '/var/lib/postgresql/data/pg_ident.conf' -> '/var/lib/postgresql/data/old/pg_ident.conf'
renamed '/var/lib/postgresql/data/pg_logical' -> '/var/lib/postgresql/data/old/pg_logical'
renamed '/var/lib/postgresql/data/pg_multixact' -> '/var/lib/postgresql/data/old/pg_multixact'
renamed '/var/lib/postgresql/data/pg_notify' -> '/var/lib/postgresql/data/old/pg_notify'
renamed '/var/lib/postgresql/data/pg_replslot' -> '/var/lib/postgresql/data/old/pg_replslot'
renamed '/var/lib/postgresql/data/pgroonga.log' -> '/var/lib/postgresql/data/old/pgroonga.log'
renamed '/var/lib/postgresql/data/pg_serial' -> '/var/lib/postgresql/data/old/pg_serial'
renamed '/var/lib/postgresql/data/pg_snapshots' -> '/var/lib/postgresql/data/old/pg_snapshots'
renamed '/var/lib/postgresql/data/pg_stat' -> '/var/lib/postgresql/data/old/pg_stat'
renamed '/var/lib/postgresql/data/pg_stat_tmp' -> '/var/lib/postgresql/data/old/pg_stat_tmp'
renamed '/var/lib/postgresql/data/pg_subtrans' -> '/var/lib/postgresql/data/old/pg_subtrans'
renamed '/var/lib/postgresql/data/pg_tblspc' -> '/var/lib/postgresql/data/old/pg_tblspc'
renamed '/var/lib/postgresql/data/pg_twophase' -> '/var/lib/postgresql/data/old/pg_twophase'
renamed '/var/lib/postgresql/data/PG_VERSION' -> '/var/lib/postgresql/data/old/PG_VERSION'
renamed '/var/lib/postgresql/data/pg_wal' -> '/var/lib/postgresql/data/old/pg_wal'
renamed '/var/lib/postgresql/data/pg_xact' -> '/var/lib/postgresql/data/old/pg_xact'
renamed '/var/lib/postgresql/data/postgresql.auto.conf' -> '/var/lib/postgresql/data/old/postgresql.auto.conf'
renamed '/var/lib/postgresql/data/postgresql.conf' -> '/var/lib/postgresql/data/old/postgresql.conf'
renamed '/var/lib/postgresql/data/postmaster.opts' -> '/var/lib/postgresql/data/old/postmaster.opts'
renamed '/var/lib/postgresql/data/postmaster.pid' -> '/var/lib/postgresql/data/old/postmaster.pid'
renamed '/var/lib/postgresql/data/upgrade_in_progress.lock' -> '/var/lib/postgresql/data/old/upgrade_in_progress.lock'
-------------------------------------------------------------------
Moving existing data files into OLD temporary directory is complete
-------------------------------------------------------------------
---------------------------------------
Creating NEW temporary directory /var/lib/postgresql/data/new
---------------------------------------
--------------------------------------------
Creating NEW temporary directory is complete
--------------------------------------------
-----------------------------------------------------
Changing permissions of temporary directories to 0700
-----------------------------------------------------
---------------------------------------------------------
Changing permissions of temporary directories is complete
---------------------------------------------------------
-------------------------------------------------
Remove postmaster.pid file from PG data directory
-------------------------------------------------
------------------------------------
Determining our own initdb arguments
------------------------------------
2025-02-19 16:26:47.045 UTC [36] LOG:  database system was interrupted; last known up at 2025-02-19 16:24:33 UTC
2025-02-19 16:26:47.057 UTC [36] LOG:  database system was not properly shut down; automatic recovery in progress
2025-02-19 16:26:47.060 UTC [36] LOG:  redo starts at 0/56FB4C0
2025-02-19 16:26:47.094 UTC [36] LOG:  invalid record length at 0/5798EB0: expected at least 24, got 0
2025-02-19 16:26:47.094 UTC [36] LOG:  redo done at 0/5798E78 system usage: CPU: user: 0.00 s, system: 0.02 s, elapsed: 0.03 s
2025-02-19 16:26:47.099 UTC [36] LOG:  checkpoint starting: end-of-recovery immediate wait
2025-02-19 16:26:47.163 UTC [36] LOG:  checkpoint complete: wrote 3475 buffers (21.2%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.044 s, sync=0.016 s, total=0.066 s; sync files=27, longest=0.015 s, average=0.001 s; distance=630 kB, estimate=630 kB; lsn=0/5798EB0, redo lsn=0/5798EB0
2025-02-19 16:26:47.166 UTC [36] FATAL:  database "xd003" does not exist
2025-02-19 16:26:47.167 UTC [36] LOG:  checkpoint starting: shutdown immediate
2025-02-19 16:26:47.172 UTC [36] LOG:  checkpoint complete: wrote 0 buffers (0.0%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.001 s, sync=0.001 s, total=0.007 s; sync files=0, longest=0.000 s, average=0.000 s; distance=0 kB, estimate=567 kB; lsn=0/5798F28, redo lsn=0/5798F28
```

Docker logs of the unsuccessfully upgraded postgres container
```
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... Etc/UTC
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
syncing data to disk ... ok


Success. You can now start the database server using:

    pg_ctl -D /var/lib/postgresql/data -l logfile start

initdb: warning: enabling "trust" authentication for local connections
initdb: hint: You can change this by editing pg_hba.conf or using the option -A, or --auth-local and --auth-host, the next time you run initdb.
waiting for server to start....2025-02-19 16:24:08.418 UTC [49] LOG:  starting PostgreSQL 16.7 (Debian 16.7-1.pgdg120+1) on aarch64-unknown-linux-gnu, compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit
2025-02-19 16:24:08.419 UTC [49] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2025-02-19 16:24:08.426 UTC [52] LOG:  database system was shut down at 2025-02-19 16:24:08 UTC
2025-02-19 16:24:08.432 UTC [49] LOG:  database system is ready to accept connections
 done
server started
CREATE DATABASE


/usr/local/bin/docker-entrypoint.sh: ignoring /docker-entrypoint-initdb.d/*

waiting for server to shut down....2025-02-19 16:24:08.638 UTC [49] LOG:  received fast shutdown request
2025-02-19 16:24:08.640 UTC [49] LOG:  aborting any active transactions
2025-02-19 16:24:08.643 UTC [49] LOG:  background worker "logical replication launcher" (PID 55) exited with exit code 1
2025-02-19 16:24:08.644 UTC [50] LOG:  shutting down
2025-02-19 16:24:08.645 UTC [50] LOG:  checkpoint starting: shutdown immediate
2025-02-19 16:24:08.673 UTC [50] LOG:  checkpoint complete: wrote 922 buffers (5.6%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.017 s, sync=0.006 s, total=0.030 s; sync files=301, longest=0.003 s, average=0.001 s; distance=4255 kB, estimate=4255 kB; lsn=0/1912078, redo lsn=0/1912078
2025-02-19 16:24:08.682 UTC [49] LOG:  database system is shut down
 done
server stopped

PostgreSQL init process complete; ready for start up.

2025-02-19 16:24:08.764 UTC [1] LOG:  starting PostgreSQL 16.7 (Debian 16.7-1.pgdg120+1) on aarch64-unknown-linux-gnu, compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit
2025-02-19 16:24:08.764 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2025-02-19 16:24:08.764 UTC [1] LOG:  listening on IPv6 address "::", port 5432
2025-02-19 16:24:08.768 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2025-02-19 16:24:08.773 UTC [65] LOG:  database system was shut down at 2025-02-19 16:24:08 UTC
2025-02-19 16:24:08.780 UTC [1] LOG:  database system is ready to accept connections
2025-02-19 16:24:32.789 UTC [1] LOG:  received fast shutdown request
2025-02-19 16:24:32.791 UTC [1] LOG:  aborting any active transactions
2025-02-19 16:24:32.794 UTC [1] LOG:  background worker "logical replication launcher" (PID 68) exited with exit code 1
2025-02-19 16:24:32.797 UTC [63] LOG:  shutting down
2025-02-19 16:24:32.799 UTC [63] LOG:  checkpoint starting: shutdown immediate
2025-02-19 16:24:32.897 UTC [63] LOG:  checkpoint complete: wrote 3709 buffers (22.6%); 0 WAL file(s) added, 0 removed, 4 recycled; write=0.080 s, sync=0.009 s, total=0.101 s; sync files=138, longest=0.002 s, average=0.001 s; distance=63396 kB, estimate=63396 kB; lsn=0/56FB448, redo lsn=0/56FB448
2025-02-19 16:24:32.909 UTC [1] LOG:  database system is shut down

PostgreSQL Database directory appears to contain a database; Skipping initialization

2025-02-19 16:24:33.264 UTC [1] LOG:  starting PostgreSQL 16.7 (Debian 16.7-1.pgdg120+1) on aarch64-unknown-linux-gnu, compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit
2025-02-19 16:24:33.264 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2025-02-19 16:24:33.264 UTC [1] LOG:  listening on IPv6 address "::", port 5432
2025-02-19 16:24:33.268 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2025-02-19 16:24:33.274 UTC [29] LOG:  database system was shut down at 2025-02-19 16:24:32 UTC
2025-02-19 16:24:33.282 UTC [1] LOG:  database system is ready to accept connections
2025-02-19 16:26:53.344 UTC [39] FATAL:  could not open file "global/pg_filenode.map": No such file or directory
2025-02-19 16:27:13.344 UTC [40] FATAL:  could not open file "global/pg_filenode.map": No such file or directory
2025-02-19 16:27:25.468 UTC [1] LOG:  received fast shutdown request
2025-02-19 16:27:25.468 UTC [1] LOG:  could not open file "postmaster.pid": No such file or directory
2025-02-19 16:27:25.468 UTC [1] LOG:  aborting any active transactions
2025-02-19 16:27:25.471 UTC [1] LOG:  background worker "logical replication launcher" (PID 32) exited with exit code 1
2025-02-19 16:27:25.473 UTC [27] LOG:  shutting down
2025-02-19 16:27:25.473 UTC [27] PANIC:  could not open file "/var/lib/postgresql/data/global/pg_control": No such file or directory
2025-02-19 16:27:25.745 UTC [1] LOG:  checkpointer process (PID 27) was terminated by signal 6: Aborted
2025-02-19 16:27:25.745 UTC [1] LOG:  terminating any other active server processes
2025-02-19 16:27:25.745 UTC [1] LOG:  abnormal database system shutdown
2025-02-19 16:27:25.748 UTC [1] LOG:  database system is shut down
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
initdb: hint: If you want to create a new database system, either remove or empty the directory "/var/lib/postgresql/data" or run initdb with an argument other than "/var/lib/postgresql/data".
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.
```

**Comment by xd003** [comment]
Hello @justinclift just wanted to confirm if the issue mentioned in here https://github.com/pgautoupgrade/docker-pgautoupgrade/issues/7#issuecomment-2669177871 is the same being discussed in this thread or i should create a new one as this thread is pretty inactive

**Comment by justinclift** [comment]
@xd003 from reading over the your logs, it sounds like a new issue (😬).

I'm personally pretty time limited at the moment though, so someone else will need to take a look if they have time.  Sorry. 😦

**Comment by OdinVex** [comment]
Failure to init results in `new` being empty and everything left in `old`, breaks subsequent starts. (Edit: For instance, I run `--data-checksums --auth-host=scram-sha-256 --auth-local=scram-sha-256` for initargs, so it fusses `initdb: error: must specify a password for the superuser to enable password authentication` when I it should just be using `POSTGRES_PASSWORD`.)

**Comment by Quintasan** [comment]
I ran into this when upgrading from 14 to 15. Now my `/var/lib/docker/volumes/infisical_pg_data/_data/` has `new` and `old` inside where `new` is empty. Can I somehow retry the upgrade?

EDIT: `rmdir new && mv old/* . && rmdir old` and restarting did the trick
